packages/core/src/browser/shell/tab-bars.ts

```diff
@@ -71,18 +73,29 @@ export class TabBarRenderer extends TabBar.Renderer {
      */
     contextMenuPath?: MenuPath;
 
+    protected readonly toDispose = new DisposableCollection();
+
     // TODO refactor shell, rendered should only receive props with event handlers
     // events should be handled by clients, like ApplicationShell
     // right now it is mess: (1) client logic belong to renderer, (2) cyclic dependencies between renderes and clients
-    constructor(protected readonly contextMenuRenderer?: ContextMenuRenderer) {
+    constructor(
+        protected readonly contextMenuRenderer?: ContextMenuRenderer,
+        protected readonly decoratorService?: TabBarDecoratorService
+    ) {
         super();
+        if (this.decoratorService) {
+            this.toDispose.push(this.decoratorService);
+            this.toDispose.push(this.decoratorService.onDidChangeDecorations(() => this.tabBar && this.tabBar.update()));
+        }
     }
 
     /**
      * Render tabs with the default DOM structure, but additionally register a context
      * menu listener.
+     * @param {SideBarRenderData} data - data to render the tab.
+     * @param {boolean} decorateIcon - whether the tab icon should be decorated (e.g., with diagnostic markers).
      */
-    renderTab(data: SideBarRenderData): VirtualElement {
+    renderTab(data: SideBarRenderData, decorateIcon: boolean = true): VirtualElement {
```

Comment:
I'm not sure if `decorateIcon` should be defaulted to `true` or `false`.
I think it's more a feature to turn on than to turn off. For instance, only `editors` tabs currently have this feature enabled.

Reply:
Currently since `renderTabs` function is only present in `SideTabBar` but not in `ScrollableTabBar` (for main and bottom areas), the decorateIcon parameter has to be true by default. This `SideTabBar` > `renderTabs` can call `renderTab(data, false)` to disable decoration. The other way around is not possible right now.

Perhaps this can be changed by refactoring `ScrollableTabBar` to include `renderTabs` function. I will add a TODO comment in code to briefly explain this.